diesel/src/query_dsl/mod.rs

```diff
@@ -29,18 +35,559 @@ mod order_dsl;
+    /// # }
+    /// #
+    /// # fn main() {
+    /// #     use self::users::dsl::*;
+    /// #     let connection = establish_connection();
+    /// #     connection.execute("DELETE FROM users").unwrap();
+    /// connection.execute("INSERT INTO users (name) VALUES ('Sean'), ('Sean'), ('Sean')")
+    ///     .unwrap();
+    /// let names = users.select(name).load(&connection);
+    /// let distinct_names = users.select(name).distinct().load(&connection);
+    ///
+    /// let sean = String::from("Sean");
+    /// assert_eq!(Ok(vec![sean.clone(), sean.clone(), sean.clone()]), names);
+    /// assert_eq!(Ok(vec![sean.clone()]), distinct_names);
+    /// # }
+    /// ```
+    fn distinct(self) -> Distinct<Self>
+    where
+        Self: methods::DistinctDsl,
+    {
+        methods::DistinctDsl::distinct(self)
+    }
+
+    /// Adds the `DISTINCT ON` clause to a query.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # #[macro_use] extern crate diesel;
+    /// # include!("../doctest_setup.rs");
+    /// #
+    /// # table! {
+    /// #     users {
+    /// #         id -> Integer,
+    /// #         name -> VarChar,
+    /// #     }
+    /// # }
+    /// #
+    /// # #[derive(Queryable, Debug, PartialEq)]
+    /// # struct Animal {
+    /// #     species: String,
+    /// #     name: Option<String>,
+    /// #     legs: i32,
+    /// # }
+    /// #
+    /// # impl Animal {
+    /// #     fn new<S: Into<String>>(species: S, name: Option<&str>, legs: i32) -> Self {
+    /// #         Animal {
+    /// #             species: species.into(),
+    /// #             name: name.map(Into::into),
+    /// #             legs
+    /// #         }
+    /// #     }
+    /// # }
+    /// #
+    /// # fn main() {
+    /// #     use self::animals::dsl::*;
+    /// #     let connection = establish_connection();
+    /// #     connection.execute("DELETE FROM animals").unwrap();
+    /// connection.execute("INSERT INTO animals (species, name, legs)
+    ///                     VALUES ('dog', 'Jack', 4), ('dog', Null, 4),
+    ///                            ('spider', Null, 8)")
+    ///     .unwrap();
+    /// let all_animals = animals.select((species, name, legs)).load(&connection);
+    /// let distinct_animals = animals.select((species, name, legs)).distinct_on(species).load(&connection);
+    ///
+    /// assert_eq!(Ok(vec![Animal::new("dog", Some("Jack"), 4),
+    ///                    Animal::new("dog", None, 4),
+    ///                    Animal::new("spider", None, 8)]), all_animals);
+    /// assert_eq!(Ok(vec![Animal::new("dog", Some("Jack"), 4),
+    ///                    Animal::new("spider", None, 8)]), distinct_animals);
+    /// # }
+    /// ```
+    #[cfg(feature = "postgres")]
+    fn distinct_on<Expr>(self, expr: Expr) -> DistinctOn<Self, Expr>
+    where
+        Self: methods::DistinctOnDsl<Expr>,
+    {
+        methods::DistinctOnDsl::distinct_on(self, expr)
+    }
+
+    // FIXME: Needs usage example and doc rewrite
```

Comment:
Maybe we should open an issue for that ?

Reply:
I'm going to do a pass on this module soon.